### PR TITLE
Dynamically serve PNI images

### LINKED ISSUE #6400: Dynamically serve PNI images using Wagtail ServeView
PNI images have too many renditions to load and look for on each page request causing timeouts. 

Implement this feature: https://docs.wagtail.io/en/stable/advanced_topics/images/image_serve_view.html 

The idea is to _not_ look for all images in a single HTTP request, but to instead spread them out so each image becomes it's own HTTP request to generate a rendition on the fly in a non-blocking manner.

### PR DESCRIPTION
**Pull request by KalobTaulien**
Closes #6400
Related PRs/issues #6378 #6248 #6395 

Replaces `{% image .. %}` with `{% image_url %}` so when an http request is made to the image URL, it will dynamically create an image rendition and serve it. Caching is included, so we don't need to worry about creating multiples of the same renditions over and over again. 

**Note**: I did _not_ adjust the number of thumbnails being generated. That will be taken care of in other upcoming tickets.

### CODE REVIEW

File: network-api/networkapi/urls.py [hunk 1]
```diff
@@ -48,6 +48,11 @@
         RedirectView.as_view(url='/docs/how-do-i-wagtail/'),
         name='how-do-i-wagtail'
     ),
+    re_path(
+        r'^images/([^/]*)/(\d*)/([^/]*)/[^/]*$',
+        ServeView.as_view(action='redirect'),
+        name='wagtailimages_serve',
+    ),
```
Comment (Pomax) [inline network-api/networkapi/urls.py:55]: Small nit, but let's make this its own `image_url_tag_urls.py` file with some comments that explain that this lets us use an `image_url` tag in our templates, and what it does, in the `wagtailcustomizations` dir, and then include that patthern it exports here? Just to make it much more obvious to future devs (including ourselves) that this is a customization

**Review by Pomax — CHANGES_REQUESTED** [review]

File: network-api/networkapi/image_url_tag_urls.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+"""
+Dynamic Wagtail Images.
+https://docs.wagtail.io/en/stable/advanced_topics/images/image_serve_view.html
+
+Typically with Wagtail we use use an {% image img.object fill-XXXxYYY %} template tag
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:5]:
```suggestion
Typically with Wagtail we use an {% image img.object fill-XXXxYYY %} template tag
```

File: network-api/networkapi/image_url_tag_urls.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+to create an image rendition. These renditions are blocking processes, which is typically
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:6]:
```suggestion
to create an image rendition. Creatingh renditions is a blocking process, which is typically
```

File: network-api/networkapi/image_url_tag_urls.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+fine for a standard page with a normal amount of images. Every rendition costs one
+database query, and if the rendition doesn't exist, it needs to also create the image.
+On a page with 200 images, that's 200 extra database queries, plus rendition-creation
+time. If you have 200 images, with 10 renditions each for mobile support, that's 2000
+database queries plus rendition-creation time, which is unacceptably slow.
+
+Instead of performing all those queries up front, along with performing rendition
+creation upfront, we can use dynamic image URLs to generate an image URL and when
+that URL is requests, Wagtail will find the original image, create the rendition, and
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:15]:
```suggestion
that URL is requested, Wagtail will find the original image, create the rendition, and
```

File: network-api/networkapi/image_url_tag_urls.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+serve the image. This off loads the heavy lifting in the upfront HTTP request and defers
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:16]:
```suggestion
serve the image. This offloads the heavy lifting in the intiial HTTP request and defers
```

File: network-api/networkapi/image_url_tag_urls.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+the rendition-creation to a different HTTP request. To take advantage of this, we use
+the custom URL in this file, along with a slightly different template tag and syntax.
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:18]:
```suggestion
the custom URL declared in this file, along with a slightly different template tag and syntax.
```

**Review by Pomax — CHANGES_REQUESTED** [review]
mostly phrasing/typos =)

**Comment by KalobTaulien** [comment]
I'm not sure why exactly, but tests are failing. 

They were originally failing because there were no images attached to a ProductPage in the tests, but then I added that and got another similar error. The test class is  `networkapi.buyersguide.tests.TestBuyersGuidePage`

```
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/wagtail/images/views/serve.py", line 20, in generate_image_url                                                                                                    
    signature = generate_signature(image.id, filter_spec, key)                                                                                                                                                              
AttributeError: 'str' object has no attribute 'id' 
```

This is suggesting the `product.image` doesn't have an actual image attached to it, but I had thought I solved that already. The problem comes up in the about routable pages. The test I was specifically working with was `def about_url_test(..)` and `def test_buyersguide_about_how_to_use_view(..)` (and commented out the rest of them for simplicity)

My current thinking is that `BuyersGuidePage.get_context(...)` isn't providing the right images to the products, or just that some products are missing images. I noticed get_context() is called twice, one with a ProductPage that **has** an image, and a second time where the same ProductPage **doesn't** have an image. 

That is where I'm at right now.

File: network-api/networkapi/image_url_tag_urls.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+"""
```
Comment (Pomax) [inline network-api/networkapi/image_url_tag_urls.py:1]: let's put this in `networkapi/wagtailcustomization`?

**Comment by KalobTaulien** [comment]
Soooooo ran into a funny thing where a bunch of our buyersguide pages fail to load through tests, but load perfectly when user testing. Pomax and I both spent a chunk of time trying to resolve this issue so we can use dynamic iamge URLs but we couldnt find a solution in a timely manner. 

Instead we're going to cut out the renditions we don't need so we can at least flip the switch to fully remove cloudinary and unblock those tickets. Then we'll come back to this a bit later if we determine we still need dynamically generated image URLs.

**Review by Pomax — APPROVED** [review]

**Pull request merged**